ARROW-9054: [C++] Add ScalarAggregateOptions

rok opened the pull request:
[ARROW-9054](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-9054).

This adds `ScalarAggregateOptions` to control null behavior of `mean` and `sum` kernels.

rok commented:
As per comments in Jiras this should introduce the following:
* `sum` - don't promote type (e.g. `int32` to `int64`) if possible
* `sum` / `mean` - should have `min_count` and return `NA` if the count of non-nulls is lower than `min_count`.
* `sum` / `mean` - should short circuit when it becomes impossible for non-null count to be below `min_count`.
* `mean` - should have `skipna` and `keepna` options.

nealrichardson commented:
Should MinMax be updated to use ScalarAggregateOptions?

You should also be able to remove https://github.com/apache/arrow/blob/master/r/R/compute.R#L113-L117 and search for any other references to that jira number in the R code/tests.

rok commented:
> Should MinMax be updated to use ScalarAggregateOptions?

I'll look into it. Count would be a candidate as well.

> You should also be able to remove https://github.com/apache/arrow/blob/master/r/R/compute.R#L113-L117 and search for any other references to that jira number in the R code/tests.

Will do.

rok commented:
This is slowly coming together. Remaining todo:
* min/max must cover R edge cases
* sum - don't promote type (e.g. int32 to int64) if possible
* sum / mean - should short circuit when it becomes impossible for non-null count to be below min_count.

jorisvandenbossche commented:
> sum - don't promote type (e.g. int32 to int64) if possible

Is this desired behaviour? I think in general the resulting data type should be known based on the input types and operation (and not depend on the values)

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -37,43 +37,25 @@ class ExecContext;
 /// By default, null values are ignored
-struct ARROW_EXPORT MinMaxOptions : public FunctionOptions {
+struct ARROW_EXPORT ScalarAggregateOptions : public FunctionOptions {
   enum Mode {
```
Could this "Mode" get a more descriptive name? As I understand it, this is specifically about handling nulls?

jorisvandenbossche replied:
(ah, I see that the field itself below is called "null_handling", that's more descriptive)

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -37,43 +37,25 @@ class ExecContext;
-    /// Any nulls will result in null output
-    EMIT_NULL
+    /// Skip null values.
+    SKIPNA = 0,
```
Can you use the "NULL" terminilogy instead of "NA" (since "null" is what we generally use within Arrow), eg `SKIP_NULL`

rok replied:
Ok!

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -37,43 +37,25 @@ class ExecContext;
+    /// Skip null values.
+    SKIPNA = 0,
+    /// Calculate over all values.
+    KEEPNA,
```
Do we think there will come more possible options in the future? Because if not, this null_handling mode could also be a boolean skip_nulls=True/False ?

rok replied:
Indeed, I'll make it a boolean switch. That should make things easier to reason about too.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/aggregate_test.cc:
```diff
@@ -399,6 +498,13 @@ TYPED_TEST_SUITE(TestMeanKernelNumeric, NumericArrowTypes);
+  const ScalarAggregateOptions& options =
+      ScalarAggregateOptions(ScalarAggregateOptions::SKIPNA, 0);
+
+  ValidateMean<TypeParam>("[]", Datum(std::make_shared<ScalarType>(0)), options);
```
Just to make sure I am reading this one correctly, is it basically testing `mean([], min_count=0) == 0` ?

I am not sure that should give 0 as result? I would have expected either NaN or Null. 
Basically mean of empty is dividing 0/0, which is NaN.

rok replied:
Agreed. I think I've put it there to have it documented while I was changing other things.

jorisvandenbossche replied:
It seems you changed it to return null now? Should it rather return NaN? (I assume it could be consistent with a manual mean from sum/count with `min_count=0`)

jorisvandenbossche commented:
Looking forward to this! Added a few comments on naming and exact expected semantics.

rok commented:
This requires some more work on C Glib but is ready for review otherwise.

rok commented:
@jorisvandenbossche & @nealrichardson could you take a look at this proposal?
I'm not very confident about the c_glib and Ruby changes, who could review that?

rok commented:
Should we introduce this in `All` and `Any` as well?

kou commented:
@rok I can review c_glib and Ruby parts. I'll push changes for them this branch directly.

rok commented on r/R/compute.R:
```diff
@@ -93,30 +93,36 @@ list_compute_functions <- function(pattern = NULL, ...) {
   }
-
-  call_function(FUN, a, options = list(na.rm = na.rm))
+  if (FUN == "min_max" && na.rm && a$null_count == length(a)) {
+    # Arrow sum/mean returns with the output type equal to input. But here R expects +/-Inf for boolean
+    # and integer types which arrow will only return with float input.
+    a <- a$cast(float64())
+  }
```
Is there some better way to do this?

rok commented:
> @rok I can review c_glib and Ruby parts. I'll push changes for them to this branch directly.

Thanks @kou that looks much better! :)

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -37,43 +37,17 @@ class ExecContext;
-
-  enum Mode null_handling;
+  bool skip_nulls = true;
+  uint32_t min_count = 1;
```
Was there any former discussion on what would be the best default here? (eg both pandas and R would set this to min_count=0, I think) 
Not that it matters too much as long as it's an option to change.

rok replied:
There was no discussion about this that I noticed and you're right 0 seems the better default here. Will change.

jorisvandenbossche replied:
On the other hand, for mean/min/max, a `min_count=1` might be a more sensible default, to not get NaN/Inf/-Inf for empty/all-null arrays.

(anyway, I don't have a strong opinion on the default, and it probably greatly depends on your exact use case what you find the best default. So as long as the option is there and clearly documented, that should be fine)

rok replied:
Switching back to `min_count=0` as it seems to be preferred behavior in c++ (e.g. hash aggregate tests).
Defaults in python and R can still be set differently.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:
```diff
@@ -71,41 +72,51 @@ struct SumImpl : public ScalarAggregator {
 struct MeanImpl : public SumImpl<ArrowType, SimdLevel> {
   void Finalize(KernelContext*, Datum* out) override {
-    if (this->count == 0) {
+    if (this->count == 0 || this->count < options.min_count) {
```
Should min_counts check replace `this->count == 0 `, instead of adding it?

rok replied:
Would that distinguish between `this->count == 0` and `this->count == 1`?
What check would you propose here? I might be overthinking something here :)

jorisvandenbossche replied:
> Would that distinguish between `this->count == 0` and `this->count == 1`?

That would depend on the value of `min_count` ? (for `min_count=0`, it would not distinguish both, for `min_count=1`, it would)

rok replied:
How about the case `this->count == options.min_count == 0`?

jorisvandenbossche replied:
That would then take the `else` clause, and depending on the skip_nulls option, that might result in NaN from dividing 0/0 ?

rok replied:
Ah got it. Yeah that makes sense now.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/aggregate_test.cc:
```diff
@@ -104,71 +104,122 @@ static Datum NaiveSum(const Array& array) {
-                          std::make_shared<UInt64Scalar>(3));
+                          std::make_shared<UInt64Scalar>(3), options);
+
+  const ScalarAggregateOptions& options_min_count_zero = ScalarAggregateOptions(true, 0);
```
Can you add the keyword names when constructing the option (now that the values itself are no longer descriptive), like:

```suggestion
  const ScalarAggregateOptions& options_min_count_zero = ScalarAggregateOptions(/*skip_nulls=*/true, /*min_count=*/0);
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/aggregate_basic.cc:
```diff
@@ -232,9 +238,9 @@ namespace {
+                            "ScalarAggregateOptions"};
 
 const FunctionDoc sum_doc{
     "Sum values of a numeric array", ("Null values are ignored."), {"array"}};
```
Can you update the docstring here as well? (and the same for "mean" below)

rok replied:
Done.

rok commented on cpp/src/arrow/compute/kernels/aggregate_test.cc:
```diff
@@ -104,71 +104,128 @@ static Datum NaiveSum(const Array& array) {
+  ASSERT_OK_AND_ASSIGN(Datum result, Op(array, options, nullptr));
+
+  const auto& exp = Datum(expected);
+  const auto& res = checked_pointer_cast<ScalarType>(result.scalar());
+  if (!(std::isnan((double)res->value) && std::isnan((double)expected->value))) {
+    ASSERT_TRUE(result.Equals(exp));
+  }
```
Is there a better way to check equality when `nan`s are involved?

rok commented on cpp/src/arrow/compute/kernels/aggregate_test.cc:
```diff
@@ -374,22 +476,31 @@ static Datum NaiveMean(const Array& array) {
+  using ScalarType = typename TypeTraits<OutputType>::ScalarType;
+  const auto& res = checked_pointer_cast<ScalarType>(result.scalar());
+  const auto& exp = checked_pointer_cast<ScalarType>(expected.scalar());
+  if (!(std::isnan(res->value) && std::isnan(exp->value))) {
+    DatumEqual<OutputType>::EnsureEqual(result, expected);
+  }
```
Same here.

rok commented:
@jorisvandenbossche could you take another look please? I'm sorry for the size of this thing..

rok commented:
CI issues seem unrelated.

rok commented:
@jorisvandenbossche should we ping someone else to review?

kou commented:
@rok Could you fix this Doxygen error? https://github.com/apache/arrow/pull/9758/checks?check_run_id=2389547565#step:8:2565

```text
/arrow/cpp/src/arrow/compute/api_aggregate.h:126: error: unbalanced grouping commands (warning treated as error, aborting now)
```

And could you also rebase on master? Some CI jobs should be fixed by rebasing on master.

rok commented:
> @rok Could you fix this Doxygen error? https://github.com/apache/arrow/pull/9758/checks?check_run_id=2389547565#step:8:2565
> 
> And could you also rebase on master? Some CI jobs should be fixed by rebasing on master.

Thanks for spotting that one @kou! I didn't realize I introduced doxygen an issue there. Fixed and rebased.

cyb70289 commented:
@rok , PR #10098 changes behaviour of kernel error handling. So several aggregate_basic_xxx files conflict with master branch now.
Ping me if you have trouble fixing the conflicts. I can help out.

rok commented:
Thanks for the heads up @cyb70289! I've rebased on master and things look ok locally.

rok commented:
The error is Minio/S3 test. This is ready for review.

cyb70289 commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -207,7 +207,7 @@ def test_sum_array(arrow_type):
     assert pc.sum(arr).as_py() == 10
```
Also test arrays with Nones?

rok replied:
Added.

cyb70289 commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -1092,11 +1092,8 @@ def test_strptime():
-    assert pc.count(arr, count_mode='count_non_null').as_py() == 3
-    assert pc.count(arr, count_mode='count_null').as_py() == 2
-
-    with pytest.raises(ValueError, match="'zzz' is not a valid count_mode"):
-        pc.count(arr, count_mode='zzz')
```
Why remove this?

rok replied:
It didn't make sense to keep as it's not a mode but rather a boolean variable now and string would evaluate to true.
I reintroduced a test for the integer (min_count).

cyb70289 commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -229,7 +229,7 @@ def test_sum_chunked_array(arrow_type):
     arr = pa.chunked_array((), type=arrow_type)
     assert arr.num_chunks == 0
-    assert pc.sum(arr).as_py() is None  # noqa: E711
+    assert pc.sum(arr).as_py() == 0
```
This `sum empty array to 0` is a bit surprising to me. But looks pandas is doing the same thing.

rok replied:
Reintroduced.

cyb70289 commented on cpp/src/arrow/compute/kernels/hash_aggregate_test.cc:
```diff
@@ -659,18 +659,17 @@ TEST(GroupBy, ConcreteCaseWithValidateGroupBy) {
 
   using internal::Aggregate;
   for (auto agg : {
-           Aggregate{"hash_sum", nullptr},
```
Why `nullptr` is not accepted now?

rok replied:
Reintroduced.

cyb70289 commented on cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:
```diff
@@ -256,8 +268,10 @@ struct MinMaxImpl : public ScalarAggregator {
-    if (!state.has_values ||
-        (state.has_nulls && options.null_handling == MinMaxOptions::EMIT_NULL)) {
+    if (!state.has_values && options.skip_nulls && options.min_count == 0) {
+      values = {std::make_shared<ScalarType>(state.min),
+                std::make_shared<ScalarType>(state.max)};
```
Code in this `if` clause looks not necessary as they are exactly the same as code in the last `else`.

rok replied:
Indeed! :)
I've made this into a single if/else now.

rok commented on cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:
```diff
@@ -72,44 +73,54 @@ struct SumImpl : public ScalarAggregator {
       const double mean = static_cast<double>(this->sum) / this->count;
       out->value = std::make_shared<DoubleScalar>(mean);
+    } else {
+      const double mean = static_cast<double>(this->sum) / this->length;
```
Yeah good point. I was thinking about [this comment](https://issues.apache.org/jira/browse/ARROW-9054?focusedCommentId=17136725&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17136725) but it would be a bit surprising. Reverting.

rok commented:
Thanks for the feedback @cyb70289! I've addressed it. Could you please take another look?

cyb70289 commented on cpp/src/arrow/compute/kernels/aggregate_basic_internal.h:
```diff
@@ -256,8 +262,8 @@ struct MinMaxImpl : public ScalarAggregator {
     std::vector<std::shared_ptr<Scalar>> values;
-    if (!state.has_values ||
-        (state.has_nulls && options.null_handling == MinMaxOptions::EMIT_NULL)) {
+    if (options.min_count > 0 &&
```
Current code returns maximal numerical limit for `min([], min_count=0)`, this is strange.
I think `min_count = 0` is not a valid case for min_max, we can verify the options and fail if it's 0.
Further more, `min_count` here is only checked for 0 or non 0, so basically it's useless if 0 is invalid. We can simply ignore it. Or improve min_max kernel (count valid values) as sum kernel does.

rok replied:
This is to satisfy [expected behavior in R](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Extremes.html): *The minimum and maximum of a numeric empty set are +Inf and -Inf (in this order!) which ensures transitivity, e.g., min(x1, min(x2)) == min(x1, x2).*
Perhaps it's better to fix this in R with a simple check.

cyb70289 replied:
Makes sense. Thank you.

cyb70289 commented on cpp/src/arrow/compute/kernels/aggregate_basic.cc:
```diff
@@ -75,48 +75,55 @@ struct CountImpl : public ScalarAggregator {
-        break;
-      default:
-        return Status::Invalid("Unknown CountOptions encountered");
+    if (state.options.skip_nulls) {
+      *out = Datum(state.non_nulls);
+    } else {
+      *out = Datum(state.nulls);
```
`skip_nulls = true` matches `COUNT_NON_NULL`, the meaning is clear.

`skip_nulls = false` is a bit confusing. E.g., `count(list, skip_nulls=False)` should return the total length including nulls, but it actually returns only #nulls.
I'm okay to accept current code, just wondering if there's better way.

rok replied:
If `list.length` is not available it (`count(list, skip_nulls=False) - count(list, skip_nulls=True)`) does mean executing `count` kernel twice.

cyb70289 replied:
Yes. But for current kernel, you have to call it twice if you want the array length. So looks fair.

rok replied:
Ok, I'm pro `count(list, skip_nulls=False) == length_of_list` then. Let's see what @jorisvandenbossche thinks.

rok replied:
Actually if we do change to `count(list, skip_nulls=False) == length_of_list` option then we break with the [NaiveGroupBy comparison](https://github.com/apache/arrow/blob/ce2861713472818eea264957de4cc83d5a2c567c/cpp/src/arrow/compute/kernels/hash_aggregate_test.cc#L113). So I'd actually prefer `count(list, skip_nulls=False) == null_count`.
I suppose it's less intuitive but  it's more consistent.

cyb70289 reviewed (commented):
Sorry for another two questions

rok commented:
@cyb70289 @jorisvandenbossche can I do anything to move this forward?

cyb70289 reviewed (approved):
Looks okay to me.
Just to mention that this `ScalarAggregateOptions` actually can only be applied to a limited number of functions (min_max/sum/mean/count till now). For other aggregators like Variance/Mode/Quantile, and more to come, they require special options and will have to duplicate the null/min_count behavior.

kou commented on c_glib/arrow-glib/compute.cpp:
```diff
@@ -636,60 +637,65 @@ garrow_cast_options_new(void)
+    priv->options.skip_nulls = g_value_get_boolean(value);
+    break;
+  case PROP_MIN_COUNT:
+    priv->options.skip_nulls = g_value_get_uint(value);
```
I'll fix this variable.

kou replied:
Done.

cyb70289 commented:
@nealrichardson @jorisvandenbossche Do you have other comments to this PR?

nealrichardson commented on r/R/compute.R:
```diff
@@ -110,15 +114,16 @@ max.ArrowDatum <- function(..., na.rm = FALSE) {
-    # https://issues.apache.org/jira/browse/ARROW-9054
-    return(Scalar$create(NA_real_))
+  if (FUN %in% c("mean", "sum") && !na.rm) {
+    na.min_count = length(a)
```
Could you leave a comment explaining this? It's not obvious to me what `na.min_count` does since it's not an R concept.

rok replied:
Added a suggestion.

nealrichardson commented on r/R/compute.R:
```diff
@@ -110,15 +114,16 @@ max.ArrowDatum <- function(..., na.rm = FALSE) {
+    na.min_count = length(a)
+  }
+  if (FUN == "min_max" && na.rm && a$null_count == length(a)) {
+    Array$create(data.frame(min=Inf, max=-Inf))
```
Why this special case? Is there a reason the C++ library shouldn't handle this? (Leave a comment explaining and/or linking to the JIRA issue to do this in C++)

rok replied:
I'm not sure this is the best place to put this wart :).

nealrichardson reviewed (approved):
From the R side this seems ok and we should get this resolved, though I would ideally like to see some comments explaining/justifying some of the non-intuitive `if` statements.

kou commented:
@nealrichardson Could you confirm suggestions from @rok ? Can we merge this with these suggestions?

nealrichardson commented on r/R/compute.R:
```diff
@@ -95,10 +95,14 @@ list_compute_functions <- function(pattern = NULL, ...) {
 
 #' @export
-sum.ArrowDatum <- function(..., na.rm = FALSE) scalar_aggregate("sum", ..., na.rm = na.rm)
+sum.ArrowDatum <- function(..., na.rm = FALSE, na.min_count = 0) {
+  scalar_aggregate("sum", ..., na.rm = na.rm, na.min_count = na.min_count)
+}
```
`na.min_count` will propagate through `...` if anyone passes it (but they won't).

```suggestion
sum.ArrowDatum <- function(..., na.rm = FALSE) {
  scalar_aggregate("sum", ..., na.rm = na.rm)
}
```

nealrichardson commented on r/R/compute.R:
```diff
@@ -95,10 +95,14 @@ list_compute_functions <- function(pattern = NULL, ...) {
 
 #' @export
-mean.ArrowDatum <- function(..., na.rm = FALSE) scalar_aggregate("mean", ..., na.rm = na.rm)
+mean.ArrowDatum <- function(..., na.rm = FALSE, na.min_count = 0) {
+  scalar_aggregate("mean", ..., na.rm = na.rm, na.min_count = na.min_count)
+}
```
```suggestion
mean.ArrowDatum <- function(..., na.rm = FALSE) {
  scalar_aggregate("mean", ..., na.rm = na.rm)
}
```

rok commented:
Some checks timed out. Rebased to restart the tasks.

kou commented:
The "R / AMD64 Windows RTools 35" failure was occurred on master too: https://github.com/apache/arrow/runs/2634288057

It's unrelated to this. I'll merge this.

rok commented:
Thanks for the help everyone!
